readthedocs/doc_builder/builder.py

```diff
@@ -0,0 +1,457 @@
+import os
+from collections import defaultdict
+
+import structlog
+from django.conf import settings
+from django.utils.translation import gettext_lazy as _
+
+from readthedocs.builds.constants import EXTERNAL
+from readthedocs.doc_builder.config import load_yaml_config
+from readthedocs.doc_builder.loader import get_builder_class
+from readthedocs.doc_builder.python_environments import Conda, Virtualenv
+from readthedocs.projects.exceptions import RepositoryError
+from readthedocs.projects.models import Feature
+from readthedocs.projects.signals import after_build, before_build, before_vcs
+
+log = structlog.get_logger(__name__)
+
+
+class DocumentationBuilder:
```

Comment:
In terms of a name, I think we can just something similar to build, but just a different word that isn't super overloaded. What about `BuildDirector` or `BuildManager`? Though `Manager` is already overloaded in Django.

Is the goal eventually to have other instances of this class, or for this code to be able to implement all the various build systems that users might come up with? That might help us with naming..

Reply:
From these suggestions, I think the one I like the most is `BuildDirector` since it orchestrates the build itself.

> Is the goal eventually to have other instances of this class, or for this code to be able to implement all the various build systems that users might come up with?

I don't think we will have more classes like this one since the underlying commands for each doctool do not depend on this instance. Currently, we are building Sphinx and MkDocs with the same class. In the case we will add other doctools that are not Python related, there are some names that could change, but we would just call them in a more generic way inside each method and that would be enough.

Reply:
Sounds good. Let's go with `BuildDirector` then! 💯